source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,392 @@
+      {"content-length", &Http2HeaderValidator::validateContentLengthHeader},
+  };
+
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  auto validator_it = kHeaderValidatorMap.find(key_string_view);
+  if (validator_it != kHeaderValidatorMap.end()) {
+    const auto& validator = validator_it->second;
+    return (*this.*validator)(value);
+  }
+
+  if (key_string_view.at(0) != ':') {
+    // Validate the (non-pseudo) header name
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // kHeaderValidatorMap contains every known pseudo header. If the header name starts with ":"
+    // and we don't have a validator registered in the map, then the header name is an unknown
+    // pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateResponseHeaderEntry(const HeaderString& key,
+                                                  const HeaderString& value) {
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  if (key_string_view == ":status") {
+    // Validate the :status header against the RFC valid range (100 <= status < 600)
+    return validateStatusHeader(StatusPseudoHeaderValidationMode::ValueRange, value);
+  } else if (key_string_view == "content-length") {
+    // Validate the Content-Length header
+    return validateContentLengthHeader(value);
+  } else if (key_string_view.at(0) != ':') {
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // The only valid pseudo header for responses is :status. If the header name starts with ":"
+    // and it's not ":status", then the header name is an unknown pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  // Validate the header value
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
+Http2HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap& header_map) {
+  static const absl::node_hash_set<absl::string_view> kAllowedPseudoHeadersForConnect = {
+      ":method",
+      ":authority",
+  };
+
+  static const absl::node_hash_set<absl::string_view> kAllowedPseudoHeaders = {
+      ":method", ":scheme", ":authority", ":path"};
+
+  absl::string_view path = header_map.getPathValue();
+
+  //
+  // Step 1: verify that required pseudo headers are present.
```

Comment:
I added some code to check that pseudo headers appear first. However, I think [`HeaderList`](https://github.com/envoyproxy/envoy/blob/main/source/common/http/header_map_impl.h#L181) is designed to always have pseudo headers in the front. So, I'm unable to actually test this case. Any advice?

Reply:
Bummer, I did not realize that we place pseudo header in front in header map. https://github.com/envoyproxy/envoy/blob/696180a5af17a0de62a01b41d64d0c4513146183/source/common/http/header_map_impl.h#L182

So it looks like we can't really check the order here and it will be responsibility of the codec.

@ameily sorry to go back and forth on this, I think we can remove the ordering check here, since it will always pass.